service/src/ConstConfigIOGroup.cpp

```diff
@@ -0,0 +1,435 @@
+        result += "    aggregation: " + Agg::function_to_name(it->second->agg_function) + '\n';
+        result += "    domain: " + platform_topo().domain_type_to_name(it->second->domain) + '\n';
+        result += "    iogroup: ConstConfigIOGroup";
+        return result;
+    }
+
+    std::string ConstConfigIOGroup::control_description(
+        const std::string &control_name) const
+    {
+        throw Exception("ConstConfigIOGroup::control_description: there are "
+                        "no controls supported by the ConstConfigIOGroup",
+                        GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+    }
+
+    int ConstConfigIOGroup::signal_behavior(const std::string &signal_name) const
+    {
+        if (!is_valid_signal(signal_name)) {
+            throw Exception("ConstConfigIOGroup::signal_behavior(): "
+                            "signal_name " + signal_name +
+                            " not valid for ConstConfigIOGroup",
+                            GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+        }
+        return IOGroup::M_SIGNAL_BEHAVIOR_CONSTANT;
+    }
+
+    std::string ConstConfigIOGroup::name(void) const
+    {
+        return plugin_name();
+    }
+
+    std::string ConstConfigIOGroup::plugin_name(void)
+    {
+        return M_PLUGIN_NAME;
+    }
+
+    std::unique_ptr<IOGroup> ConstConfigIOGroup::make_plugin(void)
+    {
+        return geopm::make_unique<ConstConfigIOGroup>();
+    }
+
+    void ConstConfigIOGroup::save_control(const std::string &save_path)
+    {
+    }
+
+    void ConstConfigIOGroup::restore_control(const std::string &save_path)
+    {
+    }
+
+    void ConstConfigIOGroup::parse_config_json(const std::string &config)
+    {
+        Json root = construct_config_json_obj(config);
+
+        auto signals = root.object_items();
+        for (const auto &signal : signals) {
+            check_json_signal(signal);
+            std::string name = M_SIGNAL_PREFIX + signal.first;
+
+            auto properties = signal.second.object_items();
+            int units = IOGroup::string_to_units(
+                    properties["units"].string_value());
+            int domain_type = PlatformTopo::domain_name_to_type(
+                    properties["domain"].string_value());
+            auto agg_func = Agg::name_to_function(
+                    properties["aggregation"].string_value());
+
+            auto json_values = properties["values"].array_items();
+            std::vector<double> values;
+            for (const auto &val : json_values) {
+                if (!val.is_number()) {
+                    throw Exception("ConstConfigIOGroup::parse_config_json():"
+                                    " incorrect type for property: "
+                                    "\"values\"", GEOPM_ERROR_INVALID,
+                                    __FILE__, __LINE__);
+                }
+                values.push_back(val.number_value());
+            }
+
```

Comment:
Add check that length of array matches topo.nun_domain() for the system.  This will avoid a late failure that would occur when a user tries to read the value.